aptos-move/framework/aptos-framework/sources/stake.move

```diff
@@ -1042,114 +1065,194 @@ module aptos_framework::stake {
+                    invariant 0 <= validator_index && validator_index <= vlen;
+                    invariant vlen == len(validator_set.active_validators);
+                    invariant forall i in 0..validator_index:
+                        global<ValidatorConfig>(validator_set.active_validators[i].addr).validator_index < validator_index;
+                    invariant len(validator_perf.validators) == validator_index;
                 };
-                stake_pool.locked_until_secs =
-                    timestamp::now_seconds() + recurring_lockup_duration_secs;
+                validator_index < vlen
+            }) {
+                let validator_info = vector::borrow_mut(&mut validator_set.active_validators, validator_index);
+                let validator_config = borrow_global_mut<ValidatorConfig>(validator_info.addr);
+                validator_config.validator_index = validator_index;
+
+                vector::push_back(&mut validator_perf.validators, IndividualValidatorPerformance {
+                    successful_proposals: 0,
+                    failed_proposals: 0,
+                });
+
+                // Automatically renew a validator's lockup for validators that will still be in the validator set in the
+                // next epoch.
+                let stake_pool = borrow_global_mut<StakePool>(validator_info.addr);
+                let now_secs = timestamp::now_seconds(); //TODO: change to reconfig start time
+                if (stake_pool.locked_until_secs <= now_secs) {
+                    spec {
+                        assume now_secs + recurring_lockup_duration_secs <= MAX_U64;
+                    };
+                    stake_pool.locked_until_secs =
+                        now_secs + recurring_lockup_duration_secs;
+                };
+
+                validator_index = validator_index + 1;
             };
 
-            validator_index = validator_index + 1;
+            if (features::periodical_reward_rate_decrease_enabled()) {
+                // Update rewards rate after reward distribution.
+                staking_config::calculate_and_save_latest_epoch_rewards_rate();
+            };
         };
 
-        if (features::periodical_reward_rate_decrease_enabled()) {
-            // Update rewards rate after reward distribution.
-            staking_config::calculate_and_save_latest_epoch_rewards_rate();
+        validator_consensus_infos_from_validator_set(&new_validator_set)
+    }
+
+    /// Return the `ValidatorConsensusInfo` of each current validator, sorted by current validator index.
+    public fun cur_validator_consensus_infos(): vector<ValidatorConsensusInfo> acquires ValidatorSet {
+        let validator_set = borrow_global<ValidatorSet>(@aptos_framework);
+        validator_consensus_infos_from_validator_set(validator_set)
+    }
+
+    fun validator_consensus_infos_from_validator_set(validator_set: &ValidatorSet): vector<ValidatorConsensusInfo> {
```

Comment:
why we don't re-use the ValidatorSet struct?

Reply:
A lot of thing in ValidatorSet is unnecessary (e.g., network address, pending info) and makes testing harder. ValidatorConsensusInfo is accurate.